### PR TITLE
feat: add billing interfaces and firehose ingestor aggregator

### PR DESCRIPTION
**Pull request by suhussai**
### Issue # (if applicable)

Closes #<issue number here>.

### Reason for this change

Add billing-interface for control plane.

### Description of changes

- added billing interface
- added Firehose ingestor and aggregator as a standalone solution
- update EventManager functionality
- fix bugs and responses for tenant management lambda
- simplify interfaces for control plane and core app plane

### Checklist

- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/awslabs/sbt-aws/blob/main/CONTRIBUTING.md)

---

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.*

### CODE REVIEW

File: .projenrc.ts [hunk 1]
```diff
@@ -20,10 +20,19 @@ const project = new awscdk.AwsCdkConstructLibrary({
...
+    '@aws-cdk/aws-kinesisfirehose-alpha',
+    '@aws-cdk/aws-kinesisfirehose-destinations-alpha',
+  ],
   description:
     'SaaS Builder Toolkit for AWS is a developer toolkit to implement SaaS best practices and increase developer velocity.',
-  devDeps: ['eslint-plugin-header'],
+  devDeps: [
+    `aws-cdk@${CDK_VERSION}`,
+    'eslint-plugin-header',
+    '@aws-cdk/aws-kinesisfirehose-alpha@2.123.0-alpha.0',
```
Comment (suhussai) [inline .projenrc.ts:34]: parameterize `2.123.0` version

File: src/control-plane/billing/billing-interface.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import { IFunction } from 'aws-cdk-lib/aws-lambda';
+import { IDataIngestorAggregator } from '../ingestor-aggregator/ingestor-aggregator-interface';
+
+export interface IBilling {
+  webhookPath: string;
+  createUserFunction: IFunction;
+  deleteUserFunction: IFunction;
+  ingestor: IDataIngestorAggregator;
+  putUsageFunction: IFunction;
+  webhookFunction: IFunction;
```
Comment (suhussai) [inline src/control-plane/billing/billing-interface.ts:13]: make webhook optional

File: src/control-plane/ingestor-aggregator/firehose-ingestor-aggregator.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+   */
+  readonly aggregateKeyPath: string;
+
+  /**
+   * The JMESPath to find the numeric value of key in the incoming data stream that will be aggregated.
+   */
+  readonly aggregateValuePath: string;
+}
+
+export class FirehoseIngestorAggregator extends Construct implements IDataIngestorAggregator {
```
Comment (suhussai) [inline src/control-plane/ingestor-aggregator/firehose-ingestor-aggregator.ts:45]: change to `FirehoseAggregator`

File: src/utils/event-manager.ts [hunk 4]
```diff
@@ -12,6 +33,29 @@ export interface EventManagerProps {
    * The event bus to register new rules with.
    */
   readonly eventBus: IEventBus;
+
+  readonly eventMetadata?: EventMetadata;
+}
+
+export enum EventManagerEvent {
```
Comment (suhussai) [inline src/utils/event-manager.ts:40]: update as an enum

File: src/utils/event-manager.ts [hunk 5]
```diff
@@ -22,29 +66,101 @@ export class EventManager extends Construct {
    * The event bus to register new rules with.
    * @attribute
    */
-  private readonly eventBus: IEventBus;
+  public readonly eventBus: IEventBus;
+
+  public readonly applicationPlaneEventSource: string = 'testApplicationPlaneEventSource';
```
Comment (tobuck-aws) [inline src/utils/event-manager.ts:71]: Do we want these to be "test..."?

File: src/utils/event-manager.ts [hunk 6]
```diff
@@ -22,29 +66,101 @@ export class EventManager extends Construct {
...
+  public readonly controlPlaneEventSource: string = 'testControlPlaneEventSource';
+  // sensible defaults so they are not required when instantiating control plane
+  readonly eventMetadata: EventMetadata = {
```
Comment (tobuck-aws) [inline src/utils/event-manager.ts:74]: See [TS Playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwDYEMDOa4CUCuThwDeAUHGXCgCaUAqKUA5sDABQz1MwBcFAdgJ4BKHigFEAvsXEBuYqEiw4AS14xgUAGYoEBAJK58dRswmz50eCrWbtBAKIA3YKoAKKGNd5FS5NBBxQOjxoMFAqDADaALoAND5klMwoSkg0-GDAwaHh0aSScuAWcDDp9k6qALLMlO4ocAC83mQRANbA-Mpejs4waRlRAPw83a7unrL55orOOAC2cCO9pU1wAPIAcgBCqwCCWAAiuusA4gD6WHYAigCqdgDKNA1wAOQQvABGEPSU4VjAAI44YAhZ5xMgbbZ7Q4nU53a4AYXh9zuT1eHy+UB+vAYdxwCB0GFBPghuwORzOADEdroADLXC6ot6fb7hCnJJABYBE8EUimQskwi43e6PRqvDQaZmY34AoEgsFrXn86FnOGI5GMiVSrE4vEEtDcxV80kq05U2n0uyayUYnVslKcw0uLCrABqujuug2sIRSLuKLFYCgEAcSjQSjeuPxwINCudbo9XvWZupdIZgeDofDb3tHKgXIV+zs8fdnu9ar9AZeiSDIbDEd4Uf1hqLJcT3vNaatYprmfrOfZjoVm1pNPJPvV-tR7xSSHCTZjhpHNLHMM7luns9Zg-zhp28JouldOxodnOV1uD1R2hgSgc7mAf0BwJge4PR5PZ4rGrFN7vD4XQkFX3Q9j1PFMLXTF4-3vNRcyHHwixAj9wKFS9RWrYAYIfJ85VfQs7GQsCv19H9MOwtRANjRDCPfYiIK7VFEgo4B4N3OJJkKRRkHQTBFgqUQUCYKAVjIIN-zUOB8yoN4kA6KA8GBHhCDgVp2iyMJsSiHgDAIcQnkIGR4ikrDKFkjo0BwMAFDUShFjQYZyhgKoYBqdhoieCJjPIIg4CZW0ZWfEJlLgPwAiCVTngQN5QggJAXFQXhgDufxAi5WI4ESdgUj6TJVMWXKADoSShcc0JFKJ9PEBUfLIFT-JZbE2LylSwrSngIiimLg3ixLktSnRngyrL2VyjqCtKYqthNcd1wuSq4GquBvPIer0Ua3VowwEK2oizrotUHqEtEfrwvSmJMqSHLSnGpyipKgVVVI-0FuqlayAAeg+hYYAQQrjKiWRjPE2CCAOpKEFs3SHLgASwAAHgmjILt0gA+J4koAd1hlAEaR4AUcU1GWEEIGfJBh84E4VYoHhaS1F0lhgDum6FhZjJhGwRTRNq-B4AU-AnhgAALMNCvB4BIeAShocKzgmfZ4BSfe5QNDgFgAEIBaVnnavIbWMeAbHGY+gAqOB3lQYWLatuBTY+5W9bISQnZMmAAi8bXZB8-JyZwS2lAQChqCMTgaAgRYFZ6ATeCE9RFkcnpcou9hjG4OB9EU0PmE5hwICUShdfIEWxep2n6cfRSo8qQThMWQRCqoWgOGYNgW5gR3yHycQgA) for suggestion. Happy to chat about this.

File: src/utils/event-manager.ts [hunk 7]
```diff
@@ -22,29 +66,101 @@ export class EventManager extends Construct {
...
+    },
+  };
+
+  private readonly rules: { [key: string]: Rule } = {};
+
   constructor(scope: Construct, id: string, props: EventManagerProps) {
     super(scope, id);
     this.eventBus = props.eventBus;
+
+    for (const [key, value] of Object.entries(this.eventMetadata)) {
```
Comment (tobuck-aws) [inline src/utils/event-manager.ts:155]: I would prefer to create rules for only those events that we've subscribed to. I might be missing something here, so let's discuss. Again, suggestion in TS Playground.

File: src/control-plane/services.ts [hunk 8]
```diff
@@ -65,18 +62,29 @@ export class Services extends Construct {
...
       runtime: Runtime.PYTHON_3_12,
       index: 'tenant_management.py',
       handler: 'lambda_handler',
       timeout: Duration.seconds(60),
       role: tenantManagementExecRole,
-      layers: [props.lambdaLayer],
+      layers: [
+        LayerVersion.fromLayerVersionArn(this, 'LambdaPowerTools', lambdaPowerToolsLayerARN),
+      ],
       environment: {
```
Comment (tobuck-aws) [inline src/control-plane/services.ts:80]: Assuming you export the enum, i.e.: 

```typescript
export enum EventType {
  ONBOARDING_REQUEST = 'onboardingRequest',
  OFFBOARDING_REQUEST = 'offboardingRequest',
  ACTIVATE_REQUEST = 'activateRequest',
  DEACTIVATE_REQUEST = 'deactivateRequest',
}
```
Then this code can just be:

```typescript
      environment: {
        EVENTBUS_NAME: props.eventManager.eventBus.eventBusName,
        EVENT_SOURCE: props.eventManager.controlPlaneEventSource,
        TENANT_DETAILS_TABLE: props.tables.tenantDetails.tableName,
        ONBOARDING_DETAIL_TYPE: EventType.ONBOARDING_REQUEST,
        OFFBOARDING_DETAIL_TYPE: EventType.OFFBOARDING_REQUEST,
        ACTIVATE_DETAIL_TYPE: EventType.ACTIVATE_REQUEST,
        DEACTIVATE_DETAIL_TYPE: EventType.DEACTIVATE_REQUEST,
      },
```

File: src/core-app-plane/bash-job-orchestrator.ts [hunk 9]
```diff
@@ -35,12 +35,13 @@ export interface BashJobOrchestratorProps extends cdk.StackProps {
   /**
    * Environment variables to import into the bash job from event details field.
    */
-  readonly importedVariables?: string[];
+  readonly environmentStringVariablesFromIncomingEvent?: string[];
```
Comment (tobuck-aws) [inline src/core-app-plane/bash-job-orchestrator.ts:38]: Like the new names.

File: src/core-app-plane/core-app-plane.ts [hunk 10]
```diff
@@ -63,15 +63,9 @@ export interface CoreApplicationPlaneJobRunnerProps {
    */
   readonly script: string;
 
-  /**
-   * The IncomingEventMetadata to use when listening for the event that will trigger this CoreApplicationPlaneJobRunner.
-   */
-  readonly incomingEvent: IncomingEventMetadata;
+  readonly incomingEvent: EventManagerEvent;
```
Comment (tobuck-aws) [inline src/core-app-plane/core-app-plane.ts:66]: Did you intentionally delete js-docs?

File: src/core-app-plane/integ.default.ts [hunk 11]
```diff
@@ -18,12 +19,8 @@ export class IntegStack extends cdk.Stack {
   constructor(scope: cdk.App, id: string, props?: IntegStackProps) {
     super(scope, id, props);
 
-    const controlPlaneSource = 'testControlPlaneEventSource';
-    const applicationPlaneSource = 'testApplicationPlaneEventSource';
-    const provisioningDetailType = 'testProvisioningDetailType';
-    const onboardingDetailType = 'Onboarding';
-    const offboardingDetailType = 'Offboarding';
-    const deprovisioningDetailType = 'testDeprovisioningDetailType';
+    const controlPlaneEventSource = 'testControlPlaneEventSource';
```
Comment (tobuck-aws) [inline src/core-app-plane/integ.default.ts:22]: Why can't these also use constants defined in `EventManager`, or elsewhere?

**Review by tobuck-aws — COMMENTED** [review]
Lots to talk about here. Let's chat when you have a moment.

File: src/control-plane/services.ts (hunk 8 above)
Comment (suhussai) [reply src/control-plane/services.ts:80]: Updated to use enum as detailType.

File: src/utils/event-manager.ts (hunk 7 above)
Comment (suhussai) [reply src/utils/event-manager.ts]: updated to use getOrCreate function

File: src/utils/event-manager.ts (hunk 6 above)
Comment (suhussai) [reply src/utils/event-manager.ts]: updated EventManager

File: src/utils/event-manager.ts [hunk 12]
```diff
@@ -1,9 +1,30 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-import { IEventBus, Rule, IRuleTarget } from 'aws-cdk-lib/aws-events';
+import { IEventBus, Rule, IRuleTarget, EventPattern } from 'aws-cdk-lib/aws-events';
 import { Construct } from 'constructs';
 
+export interface EventMetadata {
```
[the remark this answers is not in the document]
Comment (suhussai) [reply src/utils/event-manager.ts]: fixed.

File: src/utils/event-manager.ts (hunk 4 above)
Comment (suhussai) [reply src/utils/event-manager.ts]: updated as DetailType enum

File: src/core-app-plane/core-app-plane.ts (hunk 10 above)
Comment (suhussai) [reply src/core-app-plane/core-app-plane.ts]: updated js-docs

File: src/control-plane/billing/billing-template.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import { EventManager, EventManagerEvent } from '../../utils';
+
+export interface BillingProps {
+  readonly billing: IBilling;
+  readonly eventManager: EventManager;
+  readonly tenantDetailsTable: dynamodb.Table;
+  readonly controlPlaneAPIBillingResource: Resource;
+  readonly tenantIdColumn: string;
+}
+
```
[the remark this answers is not in the document]
Comment (suhussai) [reply src/control-plane/billing/billing-template.ts]: added more js-docs

File: src/control-plane/control-plane.ts [hunk 14]
```diff
@@ -6,30 +6,39 @@ import { EventBus } from 'aws-cdk-lib/aws-events';
 import { NagSuppressions } from 'cdk-nag';
 import { Construct } from 'constructs';
 import { IAuth } from './auth';
+import { IBilling } from './billing/billing-interface';
```
Comment (tobuck-aws) [inline src/control-plane/control-plane.ts:9]: Would it make sense to roll up everything in a barrel and export * from billing?
```suggestion
import { IBilling, BillingProvider } from './billing';
```

File: src/control-plane/services.ts [hunk 15]
```diff
@@ -65,18 +62,29 @@ export class Services extends Construct {
...
-        EVENTBUS_NAME: props.eventBus.eventBusName,
-        EVENT_SOURCE: props.controlPlaneEventSource,
+        EVENTBUS_NAME: props.eventManager.eventBus.eventBusName,
+        EVENT_SOURCE: props.eventManager.controlPlaneEventSource,
         TENANT_DETAILS_TABLE: props.tables.tenantDetails.tableName,
+        ONBOARDING_DETAIL_TYPE: DetailType.ONBOARDING_REQUEST,
```
Comment (tobuck-aws) [inline src/control-plane/services.ts:84]: So much cleaner!

**Review by tobuck-aws — COMMENTED** [review]
Looks good to me! Thanks, Syed!

File: src/control-plane/control-plane.ts (hunk 14 above)
Comment (suhussai) [reply src/control-plane/control-plane.ts]: Updated!

**Review by tobuck-aws — APPROVED** [review]

**Pull request merged**